### PR TITLE
[feat] - WriteRequest and ApplyMutation implementation for chunk server file service

### PR DESCRIPTION
**Pull request by micheal-o**
This change includes how the primary chunk server handles write request, and ask replica chunk servers to apply mutation after primary is done writing.

Changes:
1. Implementation for WriteRequest and ApplyMutation
2. Test
3. Add definition for ApplyMutation grpc
4. Fix checksum fields in grpc to use bytes instead of string, because checksum sequence isn't valid UTF-8
5. Verify sent data checksum for SendDataRequests.

### CODE REVIEW

File: src/server/chunk_server/chunk_server_file_service_impl.cc [hunk 1]
```diff
@@ -196,16 +319,164 @@ grpc::Status ChunkServerFileServiceImpl::ApplyMutations(
...
+
+  *reply->mutable_request() = *request;
+
+  // Ideally we'll use the serialized order of mutations.
+  // But we are currently just sending one apply mutation request.
+  auto& request_header = request->headers(0);
+
+  // Get data from cache and try to make the local write
+  WriteFileChunkReply write_reply;
+  auto status = WriteFileChunkInternal(request_header, &write_reply);
```
Comment (xicheng87) [inline src/server/chunk_server/chunk_server_file_service_impl.cc:332]: Technically we need to recycle the caches by calling Remove on the ChunkDataCacheManager, maybe leave a TODO here to remind ourselves.

File: src/server/chunk_server/chunk_server_file_service_impl.cc [hunk 2]
```diff
@@ -196,16 +319,164 @@ grpc::Status ChunkServerFileServiceImpl::ApplyMutations(
...
     grpc::ServerContext* context,
     const protos::grpc::SendChunkDataRequest* request,
     protos::grpc::SendChunkDataReply* reply) {
-  // TODO(someone): implement the GFS chunk server logic here
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "needs implementation");
+  *reply->mutable_request() = *request;
+
+  // use config mgr for chunksize
+  // Is the data size greater than the allowed chunk size
+  if (request->data().size() > 64 * gfs::common::bytesPerMb) {
```
Comment (xicheng87) [inline src/server/chunk_server/chunk_server_file_service_impl.cc:446]: Yeah, you want to use cofig_manager_ to fetch the configuration

File: src/server/chunk_server/chunk_server_file_service_impl.cc [hunk 3]
```diff
@@ -131,8 +142,120 @@ grpc::Status ChunkServerFileServiceImpl::ReadFileChunk(
...
+          auto apply_mutation_reply =
+              client->SendRequest(apply_mutation_request);
+
+          return std::pair<std::string, StatusOr<ApplyMutationsReply>>(
+              server_address, apply_mutation_reply);
+        }));
+  }
+
+  // Wait for the apply muatation replies
+  for (int replica = 0; replica < request->replica_locations_size();
```
Comment (xicheng87) [inline src/server/chunk_server/chunk_server_file_service_impl.cc:227]: Technically, we can save one rpc for the master to apply chunk write to itself directly right?

File: src/server/chunk_server/chunk_server_file_service_impl.cc [hunk 4]
```diff
@@ -131,8 +142,120 @@ grpc::Status ChunkServerFileServiceImpl::ReadFileChunk(
...
+
+      LOG(INFO) << "Received apply mutation status: " << apply_mutation_status
+                << " for file chunk: " << request_header.chunk_handle()
+                << " from chunk server: " << server_address;
+    } else {
+      LOG(ERROR) << "Apply mutation request to chunk server: " << server_address
+                 << " for file chunk: " << request_header.chunk_handle()
+                 << " failed. Status: " << apply_mutation_reply.status();
+
+      apply_mutation_status = FileChunkMutationStatus::UNKNOWN;
```
Comment (Michael-Tu) [inline src/server/chunk_server/chunk_server_file_service_impl.cc:248]: why not use "apply_mutation_reply.status()"?

File: src/server/chunk_server/chunk_server_file_service_impl.cc (hunk 1 above)
Comment (micheal-o) [reply src/server/chunk_server/chunk_server_file_service_impl.cc:332]: Yeah forgot to do that. Now removing cache entries after write.

File: src/server/chunk_server/chunk_server_file_service_impl.cc (hunk 3 above)
Comment (micheal-o) [reply src/server/chunk_server/chunk_server_file_service_impl.cc:227]: Hmm, you mean primary not master? My assumption was you don't include the primary server in the replica location. From the comment in the grpc: 

```
// The replicas that the client has already sent the data to, so the primary
// chunk server knows who to forward the serialized concurrent writes to.
repeated protos.ChunkServerLocation replica_locations = 2;
```

In your client request, do you also include the primary?

Comment (micheal-o) [reply src/server/chunk_server/chunk_server_file_service_impl.cc:227]: Discussed with Xi, primary isn't included in replica_locations.

**Pull request merged**